File: stockpilot-app/harmony/entry/src/main/ets/core/datasource.ts
// core/datasource.ts
// 行情数据层：东方财富为主、腾讯为备用（不同域名，可绕开对 eastmoney 的定向拦截）。
// 多候选地址自动降级 + 按域名熔断 + 内存缓存。网络通过注入的 HttpClient，可跨平台。

import type { HttpClient } from './http';
import type { Kline, Quote } from './types';

const HEADERS: Record<string, string> = {
  'User-Agent': 'Mozilla/5.0 (Linux; Android 12; Mobile) AppleWebKit/537.36',
  'Referer': 'https://quote.eastmoney.com/'
};

// 备用主机 / 协议降级表
const ALT: Record<string, string[]> = {
  'push2.eastmoney.com': ['http://push2.eastmoney.com', 'https://push2delay.eastmoney.com', 'http://push2delay.eastmoney.com'],
  'push2his.eastmoney.com': ['http://push2his.eastmoney.com'],
  'web.ifzq.gtimg.cn': ['http://web.ifzq.gtimg.cn'],
  'qt.gtimg.cn': ['http://qt.gtimg.cn']
};

interface BreakerState {
  fails: number;
  until: number;
}

interface CacheEntry {
  t: number;
  v: string;
}

const BREAKER_THRESHOLD = 3;
const BREAKER_COOLDOWN = 30000;
const RETRIABLE = /ECONNRESET|ECONNREFUSED|ETIMEDOUT|EPIPE|socket hang up|timeout|EAI_AGAIN/i;

function hostOf(url: string): string {
  const m = url.match(/^https?:\/\/([^/]+)/);
  return m ? m[1] : 'unknown';
}

function altUrls(url: string): string[] {
  const host = hostOf(url);
  const alts = ALT[host];
  if (!alts || alts.length === 0) return [];
  const out: string[] = [];
  for (const base of alts) {
    const path = url.replace(/^https?:\/\/[^/]+/, '');
    out.push(base + path);
  }
  return out;
}

function errCode(e: unknown): string {
  if (e && typeof e === 'object') {
    const o = e as { code?: string; message?: string };
    return o.code || o.message || 'unknown';
  }
  return String(e);
}

function sleep(ms: number): Promise<void> {
  return new Promise<void>((r) => setTimeout(r, ms));
}

export class DataSource {
  private http: HttpClient;
  private breaker: Map<string, BreakerState> = new Map<string, BreakerState>();
  private cache: Map<string, CacheEntry> = new Map<string, CacheEntry>();

  constructor(http: HttpClient) {
    this.http = http;
  }

  // 带降级 / 重试 / 熔断 / 缓存的取文本
  async fetchText(url: string, timeoutMs: number, cacheTtlMs: number): Promise<string> {
    const key = url;
    const hit = this.cache.get(key);
    if (hit && Date.now() - hit.t < cacheTtlMs) return hit.v;

    const candidates: string[] = [url].concat(altUrls(url));
    let lastErr = 'network failed';

    for (let ci = 0; ci < candidates.length; ci++) {
      const cand = candidates[ci];
      const host = hostOf(cand);
      const st = this.breaker.get(host) || { fails: 0, until: 0 };
      if (Date.now() < st.until) { lastErr = host + ' 熔断冷却中'; continue; }

      const attempts = ci === 0 ? 2 : 1;
      let out: string | null = null;
      for (let i = 0; i < attempts; i++) {
        try {
          const res = await this.http.get(cand, timeoutMs, HEADERS);
          out = res.text;
          break;
        } catch (e) {
          lastErr = errCode(e);
          if (i < attempts - 1 && RETRIABLE.test(lastErr)) { await sleep(250); continue; }
          break;
        }
      }
      if (out !== null) {
        this.breaker.set(host, { fails: 0, until: 0 });
        this.cache.set(key, { t: Date.now(), v: out });
        return out;
      }
      st.fails++;
      if (st.fails >= BREAKER_THRESHOLD) st.until = Date.now() + BREAKER_COOLDOWN;
      this.breaker.set(host, st);
    }
    throw new Error(lastErr);
  }

  // ---------------- 实时报价 ----------------
  async getQuotes(secids: string[]): Promise<Quote[]> {
    if (secids.length === 0) return [];
    let list: Quote[] = [];
    try {
      list = await this.emQuotes(secids);
    } catch (e) {
      list = await this.tencentQuotes(secids);
    }
    return list;
  }

  private async emQuotes(secids: string[]): Promise<Quote[]> {
    const url = 'https://push2.eastmoney.com/api/qt/ulist.np/get?fltt=2&invt=2'
      + '&fields=f1,f2,f3,f4,f5,f6,f7,f8,f10,f12,f13,f14,f15,f16,f17,f18,f22,f62,f184,f124'
      + '&secids=' + secids.join(',');
    const txt = await this.fetchText(url, 6000, 5000);
    const j = JSON.parse(txt);
    const diff = (j && j.data && j.data.diff) || [];
    const out: Quote[] = [];
    for (const d of diff) {
      if (!d || d.f2 == null || d.f2 === '-') continue;
      out.push({
        secid: String(d.f13) + '.' + String(d.f12),
        code: String(d.f12), market: String(d.f13), name: d.f14,
        price: d.f2, changePct: d.f3, change: d.f4,
        volume: d.f5, amount: d.f6, amplitude: d.f7, turnover: d.f8, volumeRatio: d.f10,
        high: d.f15, low: d.f16, open: d.f17, prevClose: d.f18,
        speed: d.f22, mainNet: d.f62, mainPct: d.f184, time: d.f124
      });
    }
    if (out.length === 0) throw new Error('empty quotes');
    return out;
  }

  private async tencentQuotes(secids: string[]): Promise<Quote[]> {
    const syms: { secid: string; sym: string }[] = [];
    for (const sid of secids) {
      const p = sid.split('.');
      syms.push({ secid: sid, sym: (p[0] === '1' ? 'sh' : 'sz') + p[1] });
    }
    const url = 'https://qt.gtimg.cn/q=' + syms.map(s => s.sym).join(',');
    const txt = await this.fetchText(url, 6000, 5000);
    const map: Record<string, Quote> = {};
    const segs = txt.split(';');
    for (const seg of segs) {
      const m = seg.match(/v_([a-z]{2}\d+)="([^"]*)"/);
      if (!m) continue;
      const f = m[2].split('~');
      if (f.length < 52) continue;
      const px = Number(f[3]);
      if (!(px > 0)) continue;
      const p2 = m[1].slice(2);
      const mkt = m[1].startsWith('sh') ? '1' : '0';
      map[m[1]] = {
        secid: mkt + '.' + p2, code: p2, market: mkt,
        price: px, prevClose: Number(f[4]), open: Number(f[5]), volume: Number(f[6]),
        change: Number(f[31]), changePct: Number(f[32]), high: Number(f[33]), low: Number(f[34]),
        amount: Number(f[37]) * 10000, turnover: Number(f[38]), amplitude: Number(f[43]),
        volumeRatio: Number(f[49]), avg: Number(f[51]), time: Date.now()
      };
    }
    const out: Quote[] = [];
    for (const item of syms) {
      const q = map[item.sym];
      if (q) out.push(q);
    }
    if (out.length === 0) throw new Error('empty tencent quotes');
    return out;
  }

  // ---------------- K 线 ----------------
  // klt: 101 日 / 102 周 / 103 月
  async getKlines(secid: string, limit: number, klt: number): Promise<Kline[]> {
    try {
      return await this.emKline(secid, limit, klt);
    } catch (e) {
      return await this.tencentKline(secid, limit, klt);
    }
  }

  private async emKline(secid: string, limit: number, klt: number): Promise<Kline[]> {
    const url = 'https://push2his.eastmoney.com/api/qt/stock/kline/get'
      + '?secid=' + secid + '&klt=' + klt + '&fqt=1&lmt=' + limit + '&end=20500101'
      + '&fields1=f1,f2,f3,f4,f5,f6&fields2=f51,f52,f53,f54,f55,f56,f57,f58,f59,f60,f61';
    const txt = await this.fetchText(url, 8000, 60000);
    const j = JSON.parse(txt);
    const kl = (j && j.data && j.data.klines) || [];
    const rows: Kline[] = [];
    for (const line of kl) {
      const p = String(line).split(',');
      const close = Number(p[2]);
      if (!(close > 0)) continue;
      rows.push({ date: p[0], open: Number(p[1]), close, high: Number(p[3]), low: Number(p[4]), volume: Number(p[5]) });
    }
    if (rows.length === 0) throw new Error('empty klines');
    return rows;
  }

  private async tencentKline(secid: string, limit: number, klt: number): Promise<Kline[]> {
    const period = klt === 102 ? 'week' : (klt === 103 ? 'month' : (klt === 101 ? 'day' : ''));
    if (!period) throw new Error('unsupported klt ' + klt);
    const p = secid.split('.');
    const sym = (p[0] === '1' ? 'sh' : 'sz') + p[1];
    const url = 'https://web.ifzq.gtimg.cn/appstock/app/fqkline/get?param=' + sym + ',' + period + ',,,' + limit + ',qfq';
    const txt = await this.fetchText(url, 8000, 60000);
    const j = JSON.parse(txt);
    const node = j && j.data && j.data[sym];
    if (!node) throw new Error('empty tencent kline');
    const arr = node['qfq' + period] || node[period] || [];
    const rows: Kline[] = [];
    for (const r of arr) {
      const close = Number(r[2]);
      if (!(close > 0)) continue;
      rows.push({ date: r[0], open: Number(r[1]), close, high: Number(r[3]), low: Number(r[4]), volume: Number(r[5]) });
    }
    if (rows.length === 0) throw new Error('empty tencent klines');
    return rows;
  }

  // ---------------- 板块 ----------------
  async getHotSectors(): Promise<{ code: string; name: string; changePct: number; mainFund: number; turnover: number }[]> {
    const url = 'https://push2.eastmoney.com/api/qt/clist/get?pn=1&pz=120&po=1&np=1&fltt=2&invt=2&fid=f3'
      + '&fs=m:90+t:3&fields=f12,f14,f3,f62,f184,f6';
    const txt = await this.fetchText(url, 8000, 60000);
    const j = JSON.parse(txt);
    const list = (j && j.data && j.data.diff) || [];
    const out: { code: string; name: string; changePct: number; mainFund: number; turnover: number }[] = [];
    for (const d of list) {
      out.push({ code: String(d.f12), name: d.f14, changePct: d.f3, mainFund: d.f62, turnover: d.f184 });
    }
    if (out.length === 0) throw new Error('empty sectors');
    return out;
  }

  async getSectorStocks(code: string): Promise<{ code: string; market: string; name: string; changePct: number; mainFund: number }[]> {
    const url = 'https://push2.eastmoney.com/api/qt/clist/get?pn=1&pz=100&po=1&np=1&fltt=2&invt=2&fid=f3'
      + '&fs=b:' + code + '&fields=f12,f13,f14,f3,f62,f184,f6';
    const txt = await this.fetchText(url, 8000, 60000);
    const j = JSON.parse(txt);
    const list = (j && j.data && j.data.diff) || [];
    const out: { code: string; market: string; name: string; changePct: number; mainFund: number }[] = [];
    for (const d of list) {
      out.push({ code: String(d.f12), market: String(d.f13), name: d.f14, changePct: d.f3, mainFund: d.f62 });
    }
    return out;
  }

  // ---------------- 搜索（代码 / 名称 / 拼音） ----------------
  async searchStocks(kw: string): Promise<{ code: string; market: string; name: string; secid: string }[]> {
    const token = 'D43BF7224FB9E8E14DAEAFBAB648853A';
    const url = 'https://searchapi.eastmoney.com/api/suggest/get?input=' + encodeURIComponent(kw)
      + '&type=14&token=' + token + '&client=PC&src=EM_quote&version=2019';
    const txt = await this.fetchText(url, 6000, 60000);
    const j = JSON.parse(txt);
    const rows = (j && j.QuotationCodeTable && j.QuotationCodeTable.Data) || [];
    const out: { code: string; market: string; name: string; secid: string }[] = [];
    for (const r of rows) {
      if (!r || !r.Code || !r.Name) continue;
      const code = String(r.Code);
      if (!/^\d/.test(code)) continue;
      const market = code.startsWith('6') ? '1' : '0';
      out.push({ code, market, name: r.Name, secid: market + '.' + code });
      if (out.length >= 15) break;
    }
    return out;
  }
}

File: stockpilot-app/android/app/app/src/main/java/com/stockpilot/app/core/DataSource.kt
package com.stockpilot.app.core

import okhttp3.OkHttpClient
import okhttp3.Request
import org.json.JSONArray
import org.json.JSONObject
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.TimeUnit

/**
 * 板块
 * 字段语义（已用真实接口核对）：
 *   changePct=f3 涨跌幅 ｜ mainFund=f62 主力净流入额 ｜ mainPct=f184 主力净占比
 *   turnover=f8 换手率   ｜ upCount=f104 上涨家数 ｜ downCount=f105 下跌家数
 *   leaderName=f128 领涨股名称 ｜ leaderCode=f140 领涨股代码
 */
data class Sector(
    val code: String,
    val name: String,
    val changePct: Double,
    val mainFund: Double,
    val turnover: Double,
    val mainPct: Double = 0.0,
    val upCount: Int = 0,
    val downCount: Int = 0,
    val leaderName: String = "",
    val leaderCode: String = ""
)

/** 成分股简表 */
data class StockBrief(
    val secid: String,
    val code: String,
    val name: String,
    val changePct: Double
)

/**
 * 行情数据层：东方财富为主、腾讯为备用（不同域名，可绕开对 eastmoney 的定向拦截）。
 * 多候选地址降级 + 按域名熔断 + 内存缓存。与 core/datasource.ts 行为一致。
 */
class DataSource {

    private val client: OkHttpClient = OkHttpClient.Builder()
        .connectTimeout(8, TimeUnit.SECONDS)
        .readTimeout(12, TimeUnit.SECONDS)
        .retryOnConnectionFailure(true)
        .build()

    private val breaker = ConcurrentHashMap<String, LongArray>()      // host -> [fails, untilMs]
    private val cache = ConcurrentHashMap<String, CacheEntry>()

    private class CacheEntry(val t: Long, val v: String)

    private val headers = mapOf(
        "User-Agent" to "Mozilla/5.0 (Linux; Android 12; Mobile) AppleWebKit/537.36",
        "Referer" to "https://quote.eastmoney.com/"
    )

    private val altHosts: Map<String, List<String>> = mapOf(
        "push2.eastmoney.com" to listOf("http://push2.eastmoney.com", "https://push2delay.eastmoney.com", "http://push2delay.eastmoney.com"),
        "push2his.eastmoney.com" to listOf("http://push2his.eastmoney.com"),
        "web.ifzq.gtimg.cn" to listOf("http://web.ifzq.gtimg.cn"),
        "qt.gtimg.cn" to listOf("http://qt.gtimg.cn")
    )

    companion object {
        private const val BREAKER_THRESHOLD = 3
        private const val BREAKER_COOLDOWN = 30_000L
    }

    private fun hostOf(url: String): String {
        val i = url.indexOf("://")
        if (i < 0) return "unknown"
        val rest = url.substring(i + 3)
        val j = rest.indexOf('/')
        return if (j < 0) rest else rest.substring(0, j)
    }

    private fun altUrls(url: String): List<String> {
        val host = hostOf(url)
        val alts = altHosts[host] ?: return emptyList()
        val path = url.removePrefix("https://$host").removePrefix("http://$host")
        return alts.map { it + path }
    }

    /** 带降级 / 熔断 / 缓存的取文本；失败抛异常 */
    fun fetchText(url: String, cacheTtlMs: Long): String {
        val hit = cache[url]
        if (hit != null && System.currentTimeMillis() - hit.t < cacheTtlMs) return hit.v

        val candidates = ArrayList<String>()
        candidates.add(url)
        candidates.addAll(altUrls(url))

        var lastErr = "network failed"
        for ((ci, cand) in candidates.withIndex()) {
            val host = hostOf(cand)
            val st = breaker[host] ?: longArrayOf(0, 0)
            if (System.currentTimeMillis() < st[1]) {
                lastErr = "$host 熔断冷却中"
                continue
            }
            val attempts = if (ci == 0) 2 else 1
            var ok: String? = null
            for (i in 0 until attempts) {
                try {
                    ok = httpGet(cand)
                    break
                } catch (e: Exception) {
                    lastErr = e.message ?: e.javaClass.simpleName
                    if (i < attempts - 1) Thread.sleep(250)
                }
            }
            if (ok != null) {
                breaker[host] = longArrayOf(0, 0)
                cache[url] = CacheEntry(System.currentTimeMillis(), ok)
                return ok
            }
            st[0] = st[0] + 1
            if (st[0] >= BREAKER_THRESHOLD) st[1] = System.currentTimeMillis() + BREAKER_COOLDOWN
            breaker[host] = st
        }
        throw RuntimeException(lastErr)
    }

    private fun httpGet(url: String): String {
        val b = Request.Builder().url(url)
        for ((k, v) in headers) b.header(k, v)
        client.newCall(b.build()).execute().use { resp ->
            if (!resp.isSuccessful) throw RuntimeException("HTTP " + resp.code)
            val body = resp.body ?: throw RuntimeException("empty body")
            return body.string()
        }
    }

    // ---------------- 实时报价 ----------------
    fun getQuotes(secids: List<String>): List<Quote> {
        if (secids.isEmpty()) return emptyList()
        return try {
            emQuotes(secids)
        } catch (e: Exception) {
            tencentQuotes(secids)
        }
    }

    private fun emQuotes(secids: List<String>): List<Quote> {
        val url = "https://push2.eastmoney.com/api/qt/ulist.np/get?fltt=2&invt=2" +
                "&fields=f1,f2,f3,f4,f5,f6,f7,f8,f10,f12,f13,f14,f15,f16,f17,f18,f22,f62,f104,f105,f106,f184,f124" +
                "&secids=" + secids.joinToString(",")
        val txt = fetchText(url, 5_000)
        val diff = JSONObject(txt).optJSONObject("data")?.optJSONArray("diff") ?: JSONArray()
        val out = ArrayList<Quote>()
        for (i in 0 until diff.length()) {
            val d = diff.optJSONObject(i) ?: continue
            val px = jd(d, "f2") ?: continue
            out.add(
                Quote(
                    secid = js(d, "f13") + "." + js(d, "f12"),
                    code = js(d, "f12") ?: "",
                    market = js(d, "f13") ?: "",
                    name = js(d, "f14"),
                    price = px,
                    changePct = jd(d, "f3"),
                    change = jd(d, "f4"),
                    volume = jd(d, "f5"),
                    amount = jd(d, "f6"),
                    amplitude = jd(d, "f7"),
                    turnover = jd(d, "f8"),
                    volumeRatio = jd(d, "f10"),
                    high = jd(d, "f15"),
                    low = jd(d, "f16"),
                    open = jd(d, "f17"),
                    prevClose = jd(d, "f18"),
                    speed = jd(d, "f22"),
                    mainNet = jd(d, "f62"),
                    mainPct = jd(d, "f184"),
                    advanceCount = jd(d, "f104")?.toInt(),
                    declineCount = jd(d, "f105")?.toInt(),
                    flatCount = jd(d, "f106")?.toInt()
                )
            )
        }
        if (out.isEmpty()) throw RuntimeException("empty quotes")
        return out
    }

    private fun tencentQuotes(secids: List<String>): List<Quote> {
        val syms = secids.map { sid ->
            val p = sid.split(".")
            (if (p[0] == "1") "sh" else "sz") + p.getOrElse(1) { "" }
        }
        val url = "https://qt.gtimg.cn/q=" + syms.joinToString(",")
        val txt = fetchText(url, 5_000)
        val map = HashMap<String, Quote>()
        for (seg in txt.split(";")) {
            val m = Regex("v_([a-z]{2}\\d+)=\"([^\"]*)\"").find(seg) ?: continue
            val sym = m.groupValues[1]
            val f = m.groupValues[2].split("~")
            if (f.size < 52) continue
            val px = f[3].toDoubleOrNull() ?: continue
            if (px <= 0) continue
            val code = sym.substring(2)
            val mkt = if (sym.startsWith("sh")) "1" else "0"
            map[sym] = Quote(
                secid = "$mkt.$code", code = code, market = mkt,
                price = px,
                prevClose = f[4].toDoubleOrNull(),
                open = f[5].toDoubleOrNull(),
                volume = f[6].toDoubleOrNull(),
                change = f[31].toDoubleOrNull(),
                changePct = f[32].toDoubleOrNull(),
                high = f[33].toDoubleOrNull(),
                low = f[34].toDoubleOrNull(),
                amount = f[37].toDoubleOrNull()?.times(10000),
                turnover = f[38].toDoubleOrNull(),
                amplitude = f[43].toDoubleOrNull(),
                volumeRatio = f[49].toDoubleOrNull(),
                avg = f[51].toDoubleOrNull()
            )
        }
        val out = ArrayList<Quote>()
        for (sym in syms) {
            map[sym]?.let { out.add(it) }
        }
        if (out.isEmpty()) throw RuntimeException("empty tencent quotes")
        return out
    }

    // ---------------- K 线 ----------------
    fun getKlines(secid: String, limit: Int, klt: Int = 101): List<Kline> {
        return try {
            emKline(secid, limit, klt)
        } catch (e: Exception) {
            tencentKline(secid, limit, klt)
        }
    }

    private fun emKline(secid: String, limit: Int, klt: Int): List<Kline> {
        val url = "https://push2his.eastmoney.com/api/qt/stock/kline/get" +
                "?secid=$secid&klt=$klt&fqt=1&lmt=$limit&end=20500101" +
                "&fields1=f1,f2,f3,f4,f5,f6&fields2=f51,f52,f53,f54,f55,f56,f57,f58,f59,f60,f61"
        val txt = fetchText(url, 60_000)
        val arr = JSONObject(txt).optJSONObject("data")?.optJSONArray("klines") ?: JSONArray()
        val rows = ArrayList<Kline>()
        for (i in 0 until arr.length()) {
            val p = arr.optString(i).split(",")
            if (p.size < 6) continue
            val close = p[2].toDoubleOrNull() ?: continue
            if (close <= 0) continue
            rows.add(Kline(p[0], p[1].toDoubleOrNull() ?: 0.0, close, p[3].toDoubleOrNull() ?: 0.0,
                p[4].toDoubleOrNull() ?: 0.0, p[5].toDoubleOrNull() ?: 0.0))
        }
        if (rows.isEmpty()) throw RuntimeException("empty klines")
        return rows
    }

    private fun tencentKline(secid: String, limit: Int, klt: Int): List<Kline> {
        val period = when (klt) {
            101 -> "day"
            102 -> "week"
            103 -> "month"
            else -> throw RuntimeException("unsupported klt $klt")
        }
        val p = secid.split(".")
        val sym = (if (p[0] == "1") "sh" else "sz") + p.getOrElse(1) { "" }
        val url = "https://web.ifzq.gtimg.cn/appstock/app/fqkline/get?param=$sym,$period,,,$limit,qfq"
        val txt = fetchText(url, 60_000)
        val node = JSONObject(txt).optJSONObject("data")?.optJSONObject(sym)
            ?: throw RuntimeException("empty tencent kline")
        val arr = node.optJSONArray("qfq$period") ?: node.optJSONArray(period) ?: JSONArray()
        val rows = ArrayList<Kline>()
        for (i in 0 until arr.length()) {
            val r = arr.optJSONArray(i) ?: continue
            val close = r.optString(2).toDoubleOrNull() ?: continue
            if (close <= 0) continue
            rows.add(Kline(r.optString(0), r.optString(1).toDoubleOrNull() ?: 0.0, close,
                r.optString(3).toDoubleOrNull() ?: 0.0, r.optString(4).toDoubleOrNull() ?: 0.0,
                r.optString(5).toDoubleOrNull() ?: 0.0))
        }
        if (rows.isEmpty()) throw RuntimeException("empty tencent klines")
        return rows
    }

    // ---------------- 板块 ----------------
    /**
     * 概念板块全量列表（东方财富共 500+ 个，接口每页上限 100，需要翻页）。
     * 只取第一页会得到「涨幅榜前 100」，用它统计涨跌家数会严重失真，因此必须取全量。
     */
    fun getHotSectors(): List<Sector> {
        val out = ArrayList<Sector>()
        val seen = HashSet<String>()
        var page = 1
        var total = Int.MAX_VALUE
        while (out.size < total && page <= 8) {
            val url = "https://push2.eastmoney.com/api/qt/clist/get?pn=$page&pz=100&po=1&np=1&fltt=2&invt=2&fid=f3" +
                    "&fs=m:90+t:3&fields=f12,f14,f3,f8,f62,f104,f105,f128,f140,f184"
            val txt = try {
                fetchText(url, 120_000)
            } catch (e: Exception) {
                break   // 部分页失败时保留已取到的数据，避免整体不可用
            }
            val data = JSONObject(txt).optJSONObject("data") ?: break
            total = data.optInt("total", 0)
            val diff = data.optJSONArray("diff") ?: break
            if (diff.length() == 0) break
            for (i in 0 until diff.length()) {
                val d = diff.optJSONObject(i) ?: continue
                val code = js(d, "f12") ?: continue
                if (!seen.add(code)) continue
                out.add(
                    Sector(
                        code = code,
                        name = js(d, "f14") ?: "",
                        changePct = jd(d, "f3") ?: 0.0,
                        mainFund = jd(d, "f62") ?: 0.0,
                        turnover = jd(d, "f8") ?: 0.0,
                        mainPct = jd(d, "f184") ?: 0.0,
                        upCount = (jd(d, "f104") ?: 0.0).toInt(),
                        downCount = (jd(d, "f105") ?: 0.0).toInt(),
                        leaderName = js(d, "f128") ?: "",
                        leaderCode = js(d, "f140") ?: ""
                    )
                )
            }
            if (diff.length() < 100) break
            page++
        }
        if (out.isEmpty()) throw RuntimeException("empty sectors")
        return out
    }

    fun getSectorStocks(code: String): List<StockBrief> {
        val url = "https://push2.eastmoney.com/api/qt/clist/get?pn=1&pz=100&po=1&np=1&fltt=2&invt=2&fid=f3" +
                "&fs=b:$code&fields=f12,f13,f14,f3"
        val txt = fetchText(url, 60_000)
        val diff = JSONObject(txt).optJSONObject("data")?.optJSONArray("diff") ?: JSONArray()
        val out = ArrayList<StockBrief>()
        for (i in 0 until diff.length()) {
            val d = diff.optJSONObject(i) ?: continue
            val c = js(d, "f12") ?: continue
            val m = js(d, "f13") ?: continue
            out.add(StockBrief("$m.$c", c, js(d, "f14") ?: c, jd(d, "f3") ?: 0.0))
        }
        return out
    }

    // ---------------- 搜索 ----------------
    fun searchStocks(kw: String): List<StockRef> {
        val token = "D43BF7224FB9E8E14DAEAFBAB648853A"
        val url = "https://searchapi.eastmoney.com/api/suggest/get?input=" +
                java.net.URLEncoder.encode(kw, "UTF-8") +
                "&type=14&token=$token&client=PC&src=EM_quote&version=2019"
        val txt = fetchText(url, 60_000)
        val rows = JSONObject(txt).optJSONObject("QuotationCodeTable")?.optJSONArray("Data") ?: JSONArray()
        val out = ArrayList<StockRef>()
        for (i in 0 until rows.length()) {
            val r = rows.optJSONObject(i) ?: continue
            val code = js(r, "Code") ?: continue
            if (!code.matches(Regex("^\\d.*"))) continue
            val market = if (code.startsWith("6")) "1" else "0"
            out.add(StockRef("$market.$code", code, js(r, "Name") ?: code))
            if (out.size >= 15) break
        }
        return out
    }

    // ---------------- JSON 取值助手 ----------------
    private fun jd(o: JSONObject, key: String): Double? {
        if (!o.has(key) || o.isNull(key)) return null
        val v = o.opt(key)
        return when (v) {
            is Number -> v.toDouble()
            is String -> {
                if (v.isEmpty() || v == "-") null else v.toDoubleOrNull()
            }
            else -> null
        }
    }

    private fun js(o: JSONObject, key: String): String? {
        if (!o.has(key) || o.isNull(key)) return null
        return o.opt(key)?.toString()
    }
}
